src/linker/Linker/MessageOrigin.cs

```diff
@@ -39,11 +39,18 @@ public MessageOrigin (ICustomAttributeProvider? provider)
 		}
 
 		public MessageOrigin (string fileName, int sourceLine = 0, int sourceColumn = 0)
+			: this (fileName, sourceLine, sourceColumn, null)
+		{
+		}
+
+		// The assembly attribute should be specified if available as it allows assigning the diagnostic
+		// to a an assembly (we group based on assembly).
+		public MessageOrigin (string fileName, int sourceLine, int sourceColumn, AssemblyDefinition? assembly)
```

Comment:
Can you make AssemblyDefinition non-nullable? I think in this advanced ctor it would make more sense.

Reply:
Unfortunately not. The case where the `null` happens here is command line specified descriptor xml. Warnings from it have file name, and line/col numbers, but no assembly. It's a valid case and the code actually handles it. The "field" test I added tests for exactly that case.